SI-7372 fix wrong insertion point for binary & linear search.

gourlaysama opened the pull request:
It should return the position the value would have if it was a part of
the sequence. Somehow even the test was wrong.

gourlaysama committed:
SI-7372 fix wrong insertion point for binary & linear search.

It should return the position the value would have if it was a part of
the sequence. Somehow even the test was wrong.

Ichoran commented on src/library/scala/collection/Searching.scala:
```diff
@@ -88,7 +88,7 @@ object Searching {
     private def binarySearch[B >: A](elem: B, from: Int, to: Int)
     (implicit ord: Ordering[B]): SearchResult = {
-      if ((to-from) == 1) InsertionPoint(from) else {
+      if ((to-from) == 1) InsertionPoint(from + 1) else {
```
Shouldn't this be `to` instead of `from+1` since we've already established that they're the same thing?  Or, alternatively, to make this less fragile to bad input (e.g. `to` less than `from`), test `to <= from+1` (but then I don't know whether `to` or `from+1` is more expected--maybe just throw an exception instead)?

Ichoran commented on src/library/scala/collection/Searching.scala:
```diff
@@ -88,7 +88,7 @@ object Searching {
     @tailrec
     private def binarySearch[B >: A](elem: B, from: Int, to: Int)
```
Not your doing, but it is really unfortunate that this `from` is an exclusive bound while in the public methods `from` is inclusive.  Really increases the chance of off-by-one errors if anyone has to come back to this.  Presumably this is why it was wrong to begin with.  Maybe you should change `from` to something else, or at least add a comment warning that it isn't the same `from`?

gourlaysama replied:
You're right, that exclusive `from` is a bit unexpected. I pushed a second commit that makes it inclusive, just like the public-facing one. This also addresses your other comment by making the `InsertionPoint` condition more readable.

lrytz commented:
@gourlaysama can you find some time to address the review feedback?

gourlaysama committed:
SI-7372 make binarySearch methods use the same parameters for the same things.

The public facing binarySearch method used an inclusive 'from'
parameter, while the internal @tailrec one used an *exclusive*
'from' parameter. No wonder there was an off-by-one error somewhere.

This makes both methods use the same exclusive 'from' parameter.

gourlaysama commented:
@lrytz thanks for the reminder, and sorry for the delay :)

adriaanm commented:
Please re-review, @Ichoran

Ichoran commented on src/library/scala/collection/Searching.scala:
```diff
@@ -81,18 +81,18 @@ object Searching {
     (implicit ord: Ordering[B]): SearchResult = {
-      if ((to-from) == 1) InsertionPoint(from) else {
+      if (to == from) InsertionPoint(from) else {
         val idx = from+(to-from)/2
```
Doesn't this change behavior when (to-from) was even and now is one less and odd (rounding down)?

gourlaysama replied:
Actually, I was hoping it would (to compensate for `from` being incremented, so that `idx` stayed the same), but, looking at it again, it obviously doesn't half the time. So for example, with this commit, searching a list of 10 elements would start by trying `idx = 5` instead of `idx = 4`, which is plain wrong. It doesn't change the result of the search, but it just isn't the middle. I was lazy and only checked the result, not the math.

To get the same values as before, now that `from` is inclusive, we need to decrement the old definition of `idx`, as in: `idx = from + (to-from-1)/2`, which give exactly the same indices than before (this time I checked). See the fixed commit.

Ichoran commented:
Not sure that the index change ended up with the same behavior.

gourlaysama commented:
@Ichoran here is the new index change, see my answer to your comment.

Ichoran commented:
LGTM
